### PR TITLE
Tweak the agent registration protocol a bit (#2464)

### PR DESCRIPTION
**Pull request by ketan**
* The server responds with a 202 (Accepted) and an empty JSON object `{}`
  as long as the registration is pending.
* The server responds with a 200 (OK) and the old JSON object containing
  the private key and the cert chain.

### CODE REVIEW

File: common/test/unit/com/thoughtworks/go/security/RegistrationTest.java [hunk 1]
```diff
@@ -39,7 +40,25 @@ public void decodeFromJson() {
...
+    @Test
+    public void shouldEncodeDecodeEmptyRegistration() throws Exception {
+        Registration toSerialize = Registration.createNullPrivateKeyEntry();
+        Registration deserialized = Registration.fromJson(toSerialize.toJson());
+
+        assertTrue(EqualsBuilder.reflectionEquals(toSerialize, deserialized));
+    }
+
+    private static Registration createRegistration() {
         File tempKeystoreFile = TestFileUtil.createUniqueTempFile(authorityKeystorePath);
```
Comment (zabil) [inline common/test/unit/com/thoughtworks/go/security/RegistrationTest.java]: Can we use junit temporary folder rule to manage temporary file creation?

File: common/test/unit/com/thoughtworks/go/security/RegistrationTest.java [hunk 2]
```diff
@@ -39,7 +40,25 @@ public void decodeFromJson() {
         assertThat(reg.getChain().length, is(3));
     }
 
-    public static Registration createRegistration() {
+
+    @Test
+    public void shouldBeValidWhenPrivateKeyAndChainIsPresent() throws Exception {
+        assertFalse(Registration.createNullPrivateKeyEntry().isValid());
```
Comment (tomzo) [inline common/test/unit/com/thoughtworks/go/security/RegistrationTest.java]: I would assert something like this too:
```
 assertFalse(Registration.fromJson("{}").isValid());
```

**Comment by arvindsv** [comment]
Does it fix [this](https://github.com/gocd/gocd/issues/2464#issuecomment-244779458) as well?

**Comment by ketan** [comment]
> Does it fix #2464 as well?

Yes, see title.

**Comment by arvindsv** [comment]
@ketan: I had linked to @tomzo's recent commit on #2464 - not just to the issue. I guess that is fixed by this as well.

**Comment by ketan** [comment]
@arvindsv — the problem looks identical, not sure why the `guid.txt` needed to be removed, something that I'd need to investigate, at first thought it looks like the culprit is [this line](https://github.com/gocd-contrib/go-cookbook/blob/master/providers/agent.rb#L66) — that file is no longer used by new agents.

It'll remain when upgrading the agents, just in case someone decides to downgrade.

@tomzo — would you be able to investigate further why the `guid.txt` needs to be removed?

**Pull request merged**